contracts/citycoin-core.clar

```diff
@@ -1,201 +1,539 @@
+)
+
+;; store user principal by user id
+(define-map Users
+  uint
+  principal
+)
+
+;; store user id by user principal
+(define-map UserIds
+  principal
+  uint
+)
+
+;; returns user ID if it exists
+(define-read-only (get-user-id (user principal))
+  (map-get? UserIds user)
+)
+
+;; returns user principal if it exists
+(define-read-only (get-user (userId uint))
+  (map-get? Users userId)
+)
+
+;; returns user ID if it has been created, or creates and returns new ID
+(define-private (get-or-create-user-id (user principal))
+  (match
+    (map-get? UserIds user)
+    value value
+    (let
+      (
+        (newId (+ u1 (var-get usersNonce)))
+      )
+      (map-set Users newId user)
+      (map-set UserIds user newId)
+      (var-set usersNonce newId)
+      newId
+    )
+  )
 )
 
-;; Allows to propose new contract that users will vote on if they want to activate it or not.
-(define-public (propose-contract (name (string-ascii 100)) (address principal))
+;; registers users that signal activation of contract until threshold is met
+(define-public (register-user (memo (optional (buff 34))))
   (let
     (
-      (newNonce (+ (var-get proposalNonce) u1))
+      (newId (+ u1 (var-get usersNonce)))
+      (threshold (var-get activationThreshold))
     )
-    (asserts! (is-authorized) (err ERR_UNAUTHORIZED))
-    (asserts! (not (is-eq (get-active-contract name) (some address))) (err ERR_CONTRACT_ALREADY_EXISTS))
-    (asserts! (is-none (get-contract address)) (err ERR_CONTRACT_ALREADY_EXISTS))
-    
-    (var-set proposalNonce newNonce)
-    (map-set Proposals 
-      newNonce
-      {
-        address: address,
-        startBH: (+ block-height u1),
-        endBH: (+ block-height u1 DEFAULT_VOTING_PERIOD),
-        voters: u0,
-        votes: u0,
-        isOpen: true
-      }
+
+    (asserts! (not (var-get initialized)) (err ERR_UNAUTHORIZED))
+
+    (asserts! (is-none (map-get? UserIds tx-sender))
+      (err ERR_USER_ALREADY_REGISTERED))
+
+    (asserts! (<= newId threshold)
+      (err ERR_ACTIVATION_THRESHOLD_REACHED))
+
+    (if (is-some memo)
+      (print memo)
+      none
+    )
+
+    (map-set Users newId tx-sender)
+    (map-set UserIds tx-sender newId)
+    (var-set usersNonce newId)
+
+    (if (is-eq newId threshold)
+      (let 
+        (
+          (activationBlockVal (+ block-height (var-get activationDelay)))
+        )
+        (var-set activationReached true)
+        (var-set activationBlock activationBlockVal)
+        (var-set coinbaseThreshold1 (+ activationBlockVal TOKEN_HALVING_BLOCKS))
+        (var-set coinbaseThreshold2 (+ activationBlockVal (* u2 TOKEN_HALVING_BLOCKS)))
+        (var-set coinbaseThreshold3 (+ activationBlockVal (* u3 TOKEN_HALVING_BLOCKS)))
+        (var-set coinbaseThreshold4 (+ activationBlockVal (* u4 TOKEN_HALVING_BLOCKS)))
+        (var-set coinbaseThreshold5 (+ activationBlockVal (* u5 TOKEN_HALVING_BLOCKS)))
+        (map-set CityCoinContracts
+          .citycoin-logic-v1
```

Comment:
You can have multiple contracts named `citycoin-logic-v1`, but they must be deployed by different wallet addresses. AFAIK it is impossible to deploy 2 contracts under the same name by one wallet.
Notation with dot at the beginning is just a sugarcoated syntax that on the backed is resolved to fully-qualified address owned by exactly the same wallet as owner of the contract in which you can find such notation.

So if we'll have 100 `citycoin-logic-v1` contracts deployed by 100 developers then no matter what `.citycoin-logic-v1` notation will be always resolved to the one deployed by us.

Reply:
Awesome! Follow-up question: what if citycoin-core is deployed by wallet X and citycoin-logic-v1 by wallet Y? How does this resolve? Or will this never happen?

Reply:
That's a very good question!
And the answer to it is "It depends".

If we would decide to perform initial deployment using multiple wallets, and we would somehow manage to perform such deployment successfully, then we would end up with unusable contracts, as they would try to call contracts that doesn't exists.

If we would perform initial deployment using only one wallet, and then upgrade `logic` to `citycoin-logic-v2`, then nothing wrong will happen, as registration of new contract is done through function to which we have to pass correct principal address.

Reply:
Nevertheless it might be a good idea to loosen up this dependency.

@whoabuddy you've made a great job implementing `setup-active-contract`. Maybe we could replace all explicit `citycoin-logic-v1` dependencies with value set by this function?
That way `core` will be depended only on `token` and `traits`

Reply:
I think this is the way to go. This refactor started out without the traits until I ran into issues in testing, so there's definitely some room to improve here.